### DISCUSSION TITLE
How to create a reviver function in Next.js to properly cast an array of strings stored in localStorage through persist into dayjs objects?

### DISCUSSION THREAD

**Discussion by anitacheung83** [pmndrs/zustand#2431] [Q&A] [answered]
Hi,
I am new to zustand have been using it for 2 weeks. Its an AMAZING state management tools.
However, I have encountered some issues in the following scenario:

**useDateStore.ts** : 
- a zustand store that stores an array of Dayjs objects
- a function that add Dayjs objects to the array

**useStore.ts**:
 - to prevent hydration error and fetch from local storage
 - from the zustand: persisting store data documentation

**page.tsx**: 
- `use client`
- contains a button to add the selected date into the array

I am trying to get the array of dates from local storage by `useStore(useDateStore, (s: any) => s.dates)`.
As the local storage will be storing `string []`, I will cast it to Dayjs by `reviver`.

Here is my code for `reviver`:

```
reviver: (key, value) => {
 if (key === "dates") {
  let dates = [];
   if (typeof value === "string" && value) {
     dates.push(dayjs(value));
   }
   return dates;
  }
  return value;
}
```

After rehydration, dates is `undefined`. I couldn't figure out why it is undefined and how to cast the string array to Dayjs array properly. 

Here is my [Code Sand Box Link](https://codesandbox.io/p/devbox/objective-browser-27n5lm?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clu1sotif00073b6jd0cp00z3%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clu1sotif00023b6j4ctoni02%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clu1sotif00043b6j8gysku62%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clu1sotif00063b6jlzcn7kcd%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B40%252C60%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clu1sotif00023b6j4ctoni02%2522%253A%257B%2522id%2522%253A%2522clu1sotif00023b6j4ctoni02%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clu1sotif00063b6jlzcn7kcd%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu1sotif00053b6jday6hhzx%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clu1sotif00063b6jlzcn7kcd%2522%252C%2522activeTabId%2522%253A%2522clu1sotif00053b6jday6hhzx%2522%257D%252C%2522clu1sotif00043b6j8gysku62%2522%253A%257B%2522id%2522%253A%2522clu1sotif00043b6j8gysku62%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu1sotif00033b6jgnve6du4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clu1sotif00033b6jgnve6du4%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

Thank you so much in advance. Would be AMAZING if someone can help me out :)

**Comment by dbritto-dev** [comment]
@anitacheung83 here's an example

```ts
reviver: (key, value: any) => {
  if (value && key === "date") {
    return parseISO(value.value);
  }
  return value;
},
replacer: (key, value) => {
  if (value && key === "date") {
    return {
      type: "date",
      value: value instanceof Date ? value.toISOString() : value,
    };
  }
  return value;
}
```

You need to combine reviver and replacer

**Reply by anitacheung83** [reply]
@dbritto-dev Thank you for your reply.
I implemented the replacer as you suggested. However, after rehydration, dates are still showing up as undefined.

Here is the [Code Sandbox link](https://codesandbox.io/p/devbox/objective-browser-27n5lm?file=%2Fapp%2Flib%2Fstores%2Fdate-store.ts%3A22%2C66&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clu1sotif00073b6jd0cp00z3%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clu1sotif00023b6j4ctoni02%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clu1sotif00043b6j8gysku62%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clu1sotif00063b6jlzcn7kcd%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B47.26384598807557%252C52.73615401192443%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clu1sotif00023b6j4ctoni02%2522%253A%257B%2522id%2522%253A%2522clu1sotif00023b6j4ctoni02%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu27xeec00023b6jseuqz1uu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A22%252C%2522startColumn%2522%253A66%252C%2522endLineNumber%2522%253A22%252C%2522endColumn%2522%253A66%257D%255D%252C%2522filepath%2522%253A%2522%252Fapp%252Flib%252Fstores%252Fdate-store.ts%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clu27xeec00023b6jseuqz1uu%2522%257D%252C%2522clu1sotif00063b6jlzcn7kcd%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu1sotif00053b6jday6hhzx%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clu1sotif00063b6jlzcn7kcd%2522%252C%2522activeTabId%2522%253A%2522clu1sotif00053b6jday6hhzx%2522%257D%252C%2522clu1sotif00043b6j8gysku62%2522%253A%257B%2522id%2522%253A%2522clu1sotif00043b6j8gysku62%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu1sotif00033b6jgnve6du4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clu1sotif00033b6jgnve6du4%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

I did double check. Unfortunately , I haven't been able to find any solution on my own. Any further guidance or suggestions would be amazing. Thank you once again.

**Reply by dbritto-dev** [reply]
@anitacheung83 Let me look into that

**Reply by dbritto-dev** [reply, marked as answer]
@anitacheung83 here you go -> 

```ts
storage: createJSONStorage(() => localStorage, {
  reviver: (key, value) => {
    if (value?.type === "date") {
      return dayjs(value.value);
    }

    return value;
  },
  replacer: (_, value) => {
    if (value instanceof dayjs) {
      return {
        type: "date",
        value: value.toJSON(),
      };
    }

    return value;
  },
}),
```
